packages/oauth-js/src/client.ts

```diff
@@ -0,0 +1,361 @@
+import Lock from "./vendor/browser-tabs-lock";
+import { AuthOptions, ClientConfig, CookieOptions } from "./types";
+import { generateBase58, hasAuthParams } from "./utils/helpers";
+import { buildAuthorizeUrl } from "./utils/urls";
+import {
+  DEFAULT_COOKIE_OPTIONS,
+  getRefreshToken,
+  getSessionData,
+  getSessionToken,
+  getSessionTokenValues,
+  getStorageAPI,
+  getTokenExpire,
+  isTokenExpiring,
+  removeTokenCookies,
+  saveSessionData,
+  setTokenCookies,
+} from "./utils/session";
+import {
+  API_VERSION,
+  LAST_PATH_KEY,
+  REFRESH_LOCK_KEY,
+  REFRESH_CHECK_INTERVAL,
+  STATE_DATA_KEY,
+  SESSION_DATA_KEY,
+  VERIFIER_DATA_KEY,
+} from "./constants";
+import { createPkceChallenge } from "./utils";
+
+export type AuthState =
+  | "INIT"
+  | "UPDATING"
+  | "AUTHENTICATED"
+  | "NOAUTH"
+  | "ERROR";
+
+export class OAuthClient {
+  config: ClientConfig;
+  options: AuthOptions;
+  authState: AuthState; // current state of authentication
+  error: string; // the last error message
+
+  private storage: Storage;
+  private timer: number | null;
+
+  constructor(config: ClientConfig, options?: CookieOptions) {
+    if (!config.clientId) throw new Error("A clientId is required");
+    if (!config.domain) throw new Error("A domain is required");
+
+    this.authState = "INIT";
+    this.error = "";
+    this.storage = getStorageAPI(!!options?.useCookie);
+    this.timer = null;
+
+    const port = ["80", "443"].includes(window.location.port)
+      ? ""
+      : `:${window.location.port}`;
+
+    this.config = {
+      ...config,
+      callbackUri:
+        config.callbackUri ||
+        `${window.location.protocol}//${window.location.hostname}${port}`,
+    };
+    this.options = {
+      sessionKey: config.sessionKey || SESSION_DATA_KEY,
+      ...DEFAULT_COOKIE_OPTIONS,
+      ...options,
+    };
+  }
+
+  async init() {
+    if (this.authState !== "INIT") {
+      return;
+    }
+
+    const tokenExpire = getTokenExpire(this.options);
+
+    if (hasAuthParams()) {
+      await this.exchange();
+    } else if (!!tokenExpire) {
+      // has a token and token will expire soon, do refresh
+      if (isTokenExpiring(tokenExpire)) {
+        try {
+          await this.refresh();
+          this.startTokenWatch();
+        } catch {}
+      } else {
+        // has a validate token
+        // TODO: call introspect to validate for opaque tokens
+        this.authState = "AUTHENTICATED";
+      }
+    } else {
+      this.authState = "NOAUTH";
+    }
+  }
+
+  async login() {
+    const { verifier, challenge } = await createPkceChallenge();
+
+    const stateCode = generateBase58(32);
+    this.storage.setItem(STATE_DATA_KEY, stateCode);
+    this.storage.setItem(VERIFIER_DATA_KEY, verifier);
+    this.storage.setItem(
+      LAST_PATH_KEY,
+      `${location.pathname}${location.search}`
+    );
+
+    const port = ["80", "443"].includes(window.location.port)
```

Comment:
is this variable used?

Reply:
not anymore, moved it to the constructor, will clean up